### PR TITLE
Rdiscrowd 4104: taskprogress api - add Null param value support

### PR DESCRIPTION
**Pull request by pwills11**
*Issue number of the reported bug or feature request: #<number>*

**Describe your changes**
A clear and concise description of the changes you have made.

**Testing performed**
Describe the testing you have performed to ensure that the bug has been addressed, or that the new feature works as planned.

**Additional context**
Add any other context about your contribution here.

### CODE REVIEW

**Review by dchhabda — DISMISSED** [review]

File: pybossa/api/__init__.py [hunk 1]
```diff
@@ -391,19 +391,20 @@ def task_progress(project_id=None, short_name=None):
...
-
     task_info_fields = get_searchable_columns(project.id)
 
     # create sql query from filter fields received on request.args
     for key in filter_fields.keys():
         if key in task_fields:
             sql_text += " AND {0}=:{1}".format(key, key)
         elif key in task_info_fields:
-            sql_text += " AND info ->> '{0}'=:{1}".format(key, key)
+            if filter_fields[key] == "null":
```
Comment (kbecker42) [inline pybossa/api/__init__.py:401]: Can `null` be case-insensitive (to support `null` and `NULL`)?

Also, we might want to consider re-using the query string for a potential optimization:

```python
sql_text +=  " AND info ->> '{0}'"

if filter_fields[key].lower() == "null":
    sql_text +=  " is Null".format(key)    
else:
    sql_text += "=:{1}".format(key, key)
```

Comment (pwills11) [reply pybossa/api/__init__.py:401]: Okay, I will change this as well. 
Good idea to use .lower() for null. I'm not sure that I understand how putting the sql_text at the top is an improvement. This implementation will break the .format() as you can see

File: pybossa/api/__init__.py [hunk 2]
```diff
@@ -384,26 +384,23 @@ def task_progress(project_id=None, short_name=None):
...
 
     sql_text = "SELECT COUNT(*) FROM task WHERE project_id=" + str(project.id)
-
     task_info_fields = get_searchable_columns(project.id)
 
     # create sql query from filter fields received on request.args
     for key in filter_fields.keys():
         if key in task_fields:
             sql_text += " AND {0}=:{1}".format(key, key)
+        elif key in task_info_fields and filter_fields[key] == "null".lower():
```
Comment (kbecker42) [inline pybossa/api/__init__.py:398]: This could possibly be optimized as below. Also `filter_fields[key].lower()` rather than "null".lower(). Thanks!

```python
elif key in task_info_fields:
  if filter_fields[key].lower() == "null":
    sql_text +=  " AND info ->> '{0}' is Null".format(key)    
  else:
    sql_text += " AND info ->> '{0}'=:{1}".format(key, key)
```

File: test/test_api/test_project_api.py [hunk 3]
```diff
@@ -1177,6 +1178,17 @@ def test_task_progress(self):
...
+        # query for the count of all tasks in the propject
+        res = self.app.get('/api/project/1/taskprogress', follow_redirects=True, headers=headers)
+        assert res.status_code == 200
+
+        # query for the count of all task count using a specific filter
+        res = self.app.get('/api/project/1/taskprogress?Fruit=apple', follow_redirects=True, headers=headers)
+        assert res.status_code == 200
+
+        # query for the count of all task count using null keyword
+        res = self.app.get('/api/project/1/taskprogress?Fruit=null', follow_redirects=True, headers=headers)
```
Comment (kbecker42) [inline test/test_api/test_project_api.py:1190]: As a sanity check, could we also add test cases for mixed-case values of `NULL` and `Null`? Thanks!

File: pybossa/api/__init__.py [hunk 4]
```diff
@@ -395,10 +395,11 @@ def task_progress(project_id=None, short_name=None):
     for key in filter_fields.keys():
         if key in task_fields:
             sql_text += " AND {0}=:{1}".format(key, key)
-        elif key in task_info_fields and filter_fields[key] == "null".lower():
-            sql_text +=  " AND info ->> '{0}' is Null".format(key)    
         elif key in task_info_fields:
-            sql_text += " AND info ->> '{0}'=:{1}".format(key, key)
+            if filter_fields[key].lower() == "null":
```
Comment (kbecker42) [inline pybossa/api/__init__.py:399]: Should the API support an empty string value?

`/api/project/1/taskprogress?Fruit=`

I'm thinking the value `null` is probably sufficient, just curious. Thanks.

File: test/test_api/test_project_api.py (hunk 3 above)
Comment (pwills11) [reply test/test_api/test_project_api.py:1190]: done!

File: pybossa/api/__init__.py (hunk 4 above)
Comment (pwills11) [reply pybossa/api/__init__.py:399]: Currently, 
/api/project/1/taskprogress?Fruit=
is not supported. 
Null is currently the supported keyword for querying empty string + null

File: test/test_api/test_project_api.py [hunk 5]
```diff
@@ -1190,6 +1190,14 @@ def test_task_progress(self):
         res = self.app.get('/api/project/1/taskprogress?Fruit=null', follow_redirects=True, headers=headers)
         assert res.status_code == 200
 
+        # query for the count of all task count using null keyword
+        res = self.app.get('/api/project/1/taskprogress?Fruit=Null', follow_redirects=True, headers=headers)
+        assert res.status_code == 200
+
+        # query for the count of all task count using null keyword
+        res = self.app.get('/api/project/1/taskprogress?Fruit=NULL', follow_redirects=True, headers=headers)
+        assert res.status_code == 200
```
Comment (kbecker42) [inline test/test_api/test_project_api.py:1199]: Should the tests assert on the expected "count" of returned tasks? Currently, it appears that only the status code is being checked, but not the actual data being returned. Thank you.

Comment (kbecker42) [reply test/test_api/test_project_api.py:1199]: It could also be useful to include a test-case for an unknown fruit type.

```python
res = self.app.get('/api/project/1/taskprogress?Fruit=INVALID', follow_redirects=True, headers=headers)
# assert the result count is 0
```

Comment (pwills11) [reply test/test_api/test_project_api.py:1199]: Okay, I can make these changes. Originally I was planning to deploy today, but at this moment I don't think that I'll be able to deploy today. Is it okay if this is delayed to Monday?

**Review by kbecker42 — DISMISSED** [review]

File: test/test_api/test_project_api.py (hunk 3 above)
Comment (dchhabda) [reply test/test_api/test_project_api.py:1190]: In order to test field with Null value and no field exist, you could add few more tasks to the tests that contains Fruits=Null and some of them without friuts field in task.info, and confirm the count value upon return in addition to  status 200.

**Review by dchhabda — APPROVED** [review]
